frame/vesting/src/lib.rs

```diff
@@ -312,105 +369,382 @@ pub mod pallet {
+			schedule.per_block(),
+			schedule.starting_block(),
+		)?;
+
+		T::Currency::transfer(
+			&source,
+			&target,
+			schedule.locked(),
+			ExistenceRequirement::AllowDeath,
+		)?;
+
+		// We can't let this fail because the currency transfer has already happened.
+		let res = Self::add_vesting_schedule(
+			&target,
+			schedule.locked(),
+			schedule.per_block(),
+			schedule.starting_block(),
+		);
+		debug_assert!(res.is_ok(), "Failed to add a schedule when we had to succeed.");
+
+		Ok(())
+	}
+
+	/// Iterate through the schedules to track the current locked amount and
+	/// filter out completed and specified schedules.
+	///
+	/// Returns a tuple that consists of:
+	/// - vec of vesting schedules, where completed schedules and those
+	/// specified by filter are remove. (Note the vec is not checked for respecting
+	/// 	bounded length.)
+	/// - the amount locked at the current block number based on the given schedules.
+	///
+	/// NOTE: the amount locked does not include any schedules that are filtered out.
+	fn report_schedule_updates(
+		schedules: Vec<VestingInfo<BalanceOf<T>, T::BlockNumber>>,
+		action: VestingAction,
+	) -> (Vec<VestingInfo<BalanceOf<T>, T::BlockNumber>>, BalanceOf<T>) {
 		let now = <frame_system::Pallet<T>>::block_number();
-		let locked_now = vesting.locked_at::<T::BlockNumberToBalance>(now);
 
-		if locked_now.is_zero() {
-			T::Currency::remove_lock(VESTING_ID, &who);
-			Vesting::<T>::remove(&who);
-			Self::deposit_event(Event::<T>::VestingCompleted(who));
+		let mut total_locked_now: BalanceOf<T> = Zero::zero();
+		let filtered_schedules = schedules
+			.into_iter()
+			.enumerate()
+			.filter_map(|(index, schedule)| {
+				let locked_now = schedule.locked_at::<T::BlockNumberToBalance>(now);
+				if locked_now.is_zero() || action.should_remove(&index) {
+					None
+				} else {
+					// We track the locked amount only if the schedule is included.
+					total_locked_now = total_locked_now.saturating_add(locked_now);
+					Some(schedule)
+				}
+			})
+			.collect::<Vec<_>>();
+
+		(filtered_schedules, total_locked_now)
+	}
+
+	/// Write an accounts updated vesting lock to storage.
+	fn write_lock(who: &T::AccountId, total_locked_now: BalanceOf<T>) {
+		if total_locked_now.is_zero() {
+			T::Currency::remove_lock(VESTING_ID, who);
+			Self::deposit_event(Event::<T>::VestingCompleted(who.clone()));
 		} else {
 			let reasons = WithdrawReasons::TRANSFER | WithdrawReasons::RESERVE;
-			T::Currency::set_lock(VESTING_ID, &who, locked_now, reasons);
-			Self::deposit_event(Event::<T>::VestingUpdated(who, locked_now));
+			T::Currency::set_lock(VESTING_ID, who, total_locked_now, reasons);
+			Self::deposit_event(Event::<T>::VestingUpdated(who.clone(), total_locked_now));
+		};
+	}
+
+	/// Write an accounts updated vesting schedules to storage.
+	fn write_vesting(
+		who: &T::AccountId,
+		schedules: Vec<VestingInfo<BalanceOf<T>, T::BlockNumber>>,
+	) -> Result<(), DispatchError> {
+		let schedules: BoundedVec<
```

Comment:
the debug assert statement about total value and number of vestings can be placed here

Reply:
Not sure I quite understand, but we can't do 
```
                debug_assert!(
                        locked_now > Zero::zero() && schedules.len() > 0 ||
                                locked_now == Zero::zero() && schedules.len() == 0
                );
 ```
 here because we don't have `locked_now`, but i did just move that check to the end of `exec_action` with my latest commit